### PR TITLE
Add option to use Clang 11's LVI Mitigation

### PR DESCRIPTION
**Pull request by asvrada**
ControlFlow mitigation only

Code Coverage won't work for clang-10, but leaving the option to compile it

Signed-off-by: Zijie Wu <zijiewu@microsoft.com>

### CODE REVIEW

File: cmake/add_enclave.cmake [hunk 1]
```diff
@@ -183,8 +184,8 @@ function (add_enclave_sgx)
   # It only re-enables the additional LVI-mitigated build of the specified enclave.
   # It does not enable the additional ctest against the LVI-mitigated version of
   # the enclave.
-  if ((LVI_MITIGATION MATCHES ControlFlow)
-      AND (ENCLAVE_ADD_LVI_MITIGATION OR NOT LVI_MITIGATION_SKIP_TESTS))
+  if ((NOT LVI_MITIGATION MATCHES None) AND (ENCLAVE_ADD_LVI_MITIGATION
```
Comment (radhikaj) [inline cmake/add_enclave.cmake:187]: Why did you change "Matches ControlFlow"

Comment (asvrada) [reply cmake/add_enclave.cmake:187]: Originally I was trying to support both ControlFlow and All option, so I changed them to check `NOT None`, to handle both ControlFlow and All.

Comment (radhikaj) [reply cmake/add_enclave.cmake:187]: Can we switch it back so the reader knows we only support ControlFlow?

**Review by radhikaj — APPROVED** [review]

**Comment by radhikaj** [comment]
Can you also do some performance analysis and add some numbers here? Thanks a lot! 
The PR looks good to me.

**Comment by asvrada** [comment]
> Can you also do some performance analysis and add some numbers here? Thanks a lot! The PR looks good to me.

Working on a performance report

File: CMakeLists.txt [hunk 2]
```diff
@@ -31,33 +31,46 @@ elseif (WIN32)
...
+  # the CMAKE_C_COMPILER cmake variable has been passed to cmake, use the C compiler
+  # that has been specified. Otherwise, prefer clang. Same for C++ compiler.
+  if (NOT DEFINED ENV{CC} AND NOT DEFINED CMAKE_C_COMPILER)
+    find_program(CMAKE_C_COMPILER clang-11 clang-10 clang)
+  endif ()
+  if (NOT DEFINED ENV{CXX} AND NOT DEFINED CMAKE_CXX_COMPILER)
+    find_program(CMAKE_CXX_COMPILER clang++-11 clang++-10 clang++)
+  endif ()
+
+  if (CMAKE_C_COMPILER MATCHES "10$" OR CMAKE_CXX_COMPILER MATCHES "10$")
```
Comment (CyanDevs) [inline CMakeLists.txt:59]: This change is good, but it seems worthwhile to point out that `CMAKE_C_COMPILER` and `CMAKE_CXX_COMPLER` ends with `clang.exe` rather than `clang-10` and thus `OE_USING_CLANG_10` won't be defined on Windows. 

It would be less confusing to have `OE_USING_CLANG_10` set according to whether clang 10 is used regardless of whether it is building on UNIX or Windows. I have some work in progress that I can adapt to address this. A preview of it is: 
https://github.com/openenclave/openenclave/commit/c29e96985cca67471eae3708f4453c8e47e3d562

Alternatively, we may want to consider logic such as `CMAKE_C_COMPILER_VERSION less than 10` or `CMAKE_C_COMPILER_VERSION greater than 10.99` so that we don't have to implement logic specific to a single version.

File: CMakeLists.txt [hunk 3]
```diff
@@ -31,33 +31,46 @@ elseif (WIN32)
...
+    set(OE_USING_CLANG_10 FALSE)
+  endif ()
+
+  # For clang-10, check options for Load Value Injection (LVI) mitigation
+  if (OE_USING_CLANG_10 AND LVI_MITIGATION MATCHES ControlFlow)
     if (${LVI_MITIGATION_BINDIR} MATCHES None)
       message(FATAL_ERROR "LVI_MITIGATION_BINDIR is not specified.")
     endif ()
+
+    # If the LVI mitigation is enabled, use the customized compilation toolchain.
```
Comment (mingweishih) [inline CMakeLists.txt:71]: Can you double check if `configure_lvi_mitigation_build` overwrite `CMAKE_C_COMPILER` and `CMAKE_CXX_COMPILER` as expected?

**Review by mingweishih — COMMENTED** [review]
Thanks for doing this. I have a high-level suggestion:
Instead of replacing the LVI mitigation with clang 11, I think we can firstly make the mitigation optional. That is, adding new options to the `LVI_MITIGATION`, for example `ControlFlow-Clang` and `ControlFlow-GNU`
When `LVI_MITIGATION=ControlFlow-Clang`, internally we check if the compiler supports it (clang version >= 11) and enable the option if the check passes.
When `LVI_MITIGATION=ControlFlow-GNU`, enable the existing GNU-based mitigation. This should be independent from the clang version.
And we can make the `ControlFlow` option equal to `ControlFlow-GNU` for now.

Doing so allows us to evaluate the runtime performance impact between two mitigations. When we think it's ready to use Clang's support, we can make `ControlFlow` default to `ControlFlow-Clang`.

Side note: there are few things to consider other than performance:
- Assembly code coverage (GNU approach covers sources written in assembly code while Clang approach may be not)
- Windows support is currently missing in this PR.

File: CMakeLists.txt (hunk 2 above)
Comment (CyanDevs) [reply CMakeLists.txt:59]: Yes, for the most part it is fine. The only part is that Windows may have issues as there are checks for `OE_USING_CLANG_10` while building in Windows, such as the case in enclave/core/CMakeLists.txt in this PR

File: .gitignore [hunk 4]
```diff
@@ -8,6 +8,8 @@
 .vscode/
 graphviz/
 build/
+build_10/
```
[the remark this answers is not in the document]
Comment (CyanDevs) [reply .gitignore:11]: Nothing creates `build_10` and `build_11` as far as I know. I'm guessing these are manually created dirs?

File: pkgconfig/CMakeLists.txt [hunk 5]
```diff
@@ -41,21 +41,31 @@ list(JOIN ENCLAVE_CFLAGS_LIST " " ENCLAVE_CFLAGS_GCC)
...
+                                 -mindirect-branch-register)
+    list(JOIN LVI_MITIGATION_FLAGS_GCC " " ENCLAVE_LVI_MITIGATION_CFLAGS_GCC)
+
+    set(LVI_MITIGATION_FLAGS_CLANG ${ENCLAVE_CFLAGS_CLANG_LIST}
+                                   ${LVI_MITIGATION_FLAGS} -no-integrated-as)
+    list(JOIN LVI_MITIGATION_FLAGS_CLANG " "
+         ENCLAVE_LVI_MITIGATION_CFLAGS_CLANG)
+  else ()
+    # clang-11 LVI
+    set(LVI_MITIGATION_FLAGS -mlvi-cfi)
```
Comment (achamayou) [inline pkgconfig/CMakeLists.txt:63]: Apologies if this is a silly question with an obvious answer, but why only control flow mitigations and not full LVI mitigations (https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-mlvi-hardening)?

There also seems to be broader option that includes LVI CFI (https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-mseses).

Comment (asvrada) [reply pkgconfig/CMakeLists.txt:63]: Thanks for reviewing! This is not official answer but I think: 

1. why only control flow mitigations and not full LVI
Currently, OE only provides custom ControlFlow LVI mitigation through a custom toolchain, so this PR is only adding ControlFlow as well.

2. why not use `-mseses`
`-mseses` turns on speculative and LVI mitigation, where currently speculative load hardening for OE is on by default, but LVI is optional. LVI comes with performance impact, so it's an option for the customer to leverage.

Comment (mingweishih) [reply pkgconfig/CMakeLists.txt:63]: @achamayou 

`SESES` approach achieves broader protection by conservatively inserting LFENCEs, which is similar to the original full LVI mitigation that Intel introduces. At the time of bringing in the LVI mitigation to OE, we decided not to support that option because of its huge performance implications.

Reference: https://reviews.llvm.org/rGbf95cf4a6816476afd76d5b981ea5fdcb9771999#change-PwDumSOD4qd4

@asvrada 

Can you also test the performance impact of the `-mlvi-hardening`? Ideally, this option should be performant while providing better coverage against LVI.

Reference:
-mlvi-hardening: https://reviews.llvm.org/D75936
-mlvi-cfi: https://reviews.llvm.org/D75934

Comment (achamayou) [reply pkgconfig/CMakeLists.txt:63]: @asvrada @mingweishih thank you for the clarification, that's very helpful.

**Review by achamayou — APPROVED** [review]

File: CMakeLists.txt (hunk 3 above)
Comment (asvrada) [reply CMakeLists.txt:74]: I tested with main branch on OpenEnclave, it doesn't seem to correctly override the compiler used. Even if this line of code is called: https://github.com/openenclave/openenclave/blob/master/cmake/configure_lvi_mitigation_build.cmake#L69

Comment (mingweishih) [reply CMakeLists.txt:74]: It might be the reason that we use if and else to separate the two cases.

Comment (asvrada) [reply CMakeLists.txt:74]:
```
SET CMAKE_C_COMPILER to ./lvi_mitigation_bin/clang-10
CMAKE_C_COMPILER: /usr/bin/clang-10
```
Even with `if else` its still reverts back to the default compiler...

Comment (mingweishih) [reply CMakeLists.txt:74]: Can you see if that works on the master branch?

Comment (asvrada) [reply CMakeLists.txt:74]: > I tested with main branch on OpenEnclave, it doesn't seem to correctly override the compiler used. Even if this line of code is called: https://github.com/openenclave/openenclave/blob/master/cmake/configure_lvi_mitigation_build.cmake#L69

Doesn't seem so on my machine.

Comment (mingweishih) [reply CMakeLists.txt:74]:
```
$ cmake .. -DLVI_MITIGATION=ControlFlow -DLVI_MITIGATION_BINDIR=/home/mingwei/lvi_mitigation_bin
-- The C compiler identification is Clang 10.0.1
-- The CXX compiler identification is Clang 10.0.1
-- The ASM compiler identification is Clang
-- Found assembler: /home/mingwei/lvi_mitigation_bin/clang-10
-- Check for working C compiler: /home/mingwei/lvi_mitigation_bin/clang-10
-- Check for working C compiler: /home/mingwei/lvi_mitigation_bin/clang-10 -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /home/mingwei/lvi_mitigation_bin/clang++-10
-- Check for working CXX compiler: /home/mingwei/lvi_mitigation_bin/clang++-10 -- works
```

Comment (CyanDevs) [reply CMakeLists.txt:76]: This was merged into the Clang-11 working branch, and I made some modifications to address the comments here. Diff: https://github.com/openenclave/openenclave/commit/f36abd814639fd9393c9b5bd250713c510d2d120#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20a

@mingweishih could you confirm the below is the expected behaviour with what you proposed? The output was obtained from running the changes in CI:

Clang 10, ControlFlow (ControlFlow-GNU)
```
+ cmake /home/oeadmin/workspace/Private/OpenEnclave/Azure-Linux -G Ninja -DCMAKE_INSTALL_PREFIX:PATH=/opt/openenclave -DCPACK_GENERATOR=DEB -DCODE_COVERAGE=OFF -DUSE_DEBUG_MALLOC=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo -DLVI_MITIGATION_BINDIR=/usr/local/lvi-mitigation/bin -DLVI_MITIGATION=ControlFlow -DLVI_MITIGATION_SKIP_TESTS=OFF -Wdev
-- The C compiler identification is Clang 10.0.0
-- The CXX compiler identification is Clang 10.0.0
-- The ASM compiler identification is Clang
-- Found assembler: /usr/local/lvi-mitigation/bin/clang-10
-- Check for working C compiler: /usr/local/lvi-mitigation/bin/clang-10
-- Check for working C compiler: /usr/local/lvi-mitigation/bin/clang-10 -- works
```
Clang 10, ControlFlow-Clang (expected failure)
```
+ cmake /home/oeadmin/workspace/Private/OpenEnclave/Azure-Linux -G Ninja -DCMAKE_INSTALL_PREFIX:PATH=/opt/openenclave -DCPACK_GENERATOR=DEB -DCODE_COVERAGE=OFF -DUSE_DEBUG_MALLOC=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo -DLVI_MITIGATION_BINDIR=/usr/local/lvi-mitigation/bin -DLVI_MITIGATION=ControlFlow-Clang -DLVI_MITIGATION_SKIP_TESTS=OFF -Wdev
-- The C compiler identification is Clang 10.0.0
-- The CXX compiler identification is Clang 10.0.0
-- The ASM compiler identification is Clang
-- Found assembler: /usr/bin/clang-10
-- Check for working C compiler: /usr/bin/clang-10
-- Check for working C compiler: /usr/bin/clang-10 -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/clang++-10
-- Check for working CXX compiler: /usr/bin/clang++-10 -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at CMakeLists.txt:95 (message):
  ControlFlow-Clang only works with clang version >= 11.  But got
  /usr/bin/clang-10
```
Clang-11, ControlFlow
```
+ cmake /home/oeadmin/workspace/Private/OpenEnclave/Azure-Linux -G Ninja -DCMAKE_INSTALL_PREFIX:PATH=/opt/openenclave -DCPACK_GENERATOR=DEB -DCODE_COVERAGE=OFF -DUSE_DEBUG_MALLOC=ON -DCMAKE_BUILD_TYPE=Debug -DLVI_MITIGATION_BINDIR=/usr/local/lvi-mitigation/bin -DLVI_MITIGATION=ControlFlow -DLVI_MITIGATION_SKIP_TESTS=ON -Wdev
-- The C compiler identification is Clang 11.1.0
-- The CXX compiler identification is Clang 11.1.0
-- The ASM compiler identification is Clang
-- Found assembler: /usr/local/lvi-mitigation/bin/clang-11
-- Check for working C compiler: /usr/local/lvi-mitigation/bin/clang-11
-- Check for working C compiler: /usr/local/lvi-mitigation/bin/clang-11 -- works
-- Detecting C compiler ABI info
```
Clang-11, ControlFlow-Clang
```
+ cmake /home/oeadmin/workspace/Private/OpenEnclave/Azure-Linux -G Ninja -DCMAKE_INSTALL_PREFIX:PATH=/opt/openenclave -DCPACK_GENERATOR=DEB -DCODE_COVERAGE=OFF -DUSE_DEBUG_MALLOC=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo -DLVI_MITIGATION_BINDIR=/usr/local/lvi-mitigation/bin -DLVI_MITIGATION=ControlFlow-Clang -DLVI_MITIGATION_SKIP_TESTS=OFF -Wdev
-- The C compiler identification is Clang 11.1.0
-- The CXX compiler identification is Clang 11.1.0
-- The ASM compiler identification is Clang
-- Found assembler: /usr/bin/clang-11
-- Check for working C compiler: /usr/bin/clang-11
-- Check for working C compiler: /usr/bin/clang-11 -- works
```

Comment (mingweishih) [reply CMakeLists.txt:76]: The behaviors are as what I expected. Thanks for doing this!
